internal/handlers/api.go

```diff
@@ -0,0 +1,402 @@
+package handlers
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi/v5"
+	"github.com/golang-jwt/jwt/v5"
+
+	"github.com/dimkene/go_final_project/internal/config"
+	"github.com/dimkene/go_final_project/internal/db/repo"
+	"github.com/dimkene/go_final_project/internal/models"
+)
+
+// это совокупность хэндлеров, часто называется api
+type Api struct {
+	repo   *repo.TasksRepository
+	config *config.Config
+}
+
+// это конструктор объекта api.
+func NewApi(repo *repo.TasksRepository, config *config.Config) *Api {
+	return &Api{repo: repo, config: config} // создаем ссылку на объект api со свойством repo, равным repo из параметров функции
+}
+
+func (a *Api) GetTaskHandler(w http.ResponseWriter, r *http.Request) {
+	idToSearch := r.URL.Query().Get("id") // это параметр запроса
+
+	// if idToSearch != "" {
+	// не нужна проверка на пустой id, потому что strconv.Atoi в таком случае всё равно вернёт ошибку
+	id, err := strconv.Atoi(idToSearch)
+	if err != nil {
+		log.Println("error:", err)
+		RenderApiErrorAndResponse(w, fmt.Errorf(InvalidIdError), http.StatusBadRequest)
+		return // иначе пойдем в a.GetTask(w, r, id) это стиль с гардами (защитниками). иначе надо написать else {a.GetTask(w, r, id)}
+	}
+	a.GetTask(w, r, id)
+}
+
+func (a *Api) PostTaskHandler(w http.ResponseWriter, r *http.Request) {
+	log.Println("We are in MethodPost")
+	a.CreateTask(w, r)
+}
+
+func (a *Api) PutTaskHandler(w http.ResponseWriter, r *http.Request) {
+	a.UpdateTask(w, r)
+}
+
+func (a *Api) DeleteTaskHandler(w http.ResponseWriter, r *http.Request) {
+	idToSearch := r.URL.Query().Get("id") // это параметр запроса
+	if idToSearch != "" {
+		a.DeleteTask(w, r)
+	} else {
+		RenderApiErrorAndResponse(w, fmt.Errorf(IdMissingError), http.StatusBadRequest)
+		return
+	}
+}
+
+// http://localhost:7540/api/tasks/257
+func (a *Api) GetTaskByIdHandler(w http.ResponseWriter, r *http.Request) {
+	idToSearch := chi.URLParam(r, "id")
+
+	id, err := strconv.Atoi(idToSearch)
+	if err != nil {
+		log.Println("error:", err)
+		RenderApiErrorAndResponse(w, fmt.Errorf(IdMissingError), http.StatusBadRequest)
+		return
+	}
+	a.GetTask(w, r, id)
+}
+
+func (a *Api) GetTasksHandler(w http.ResponseWriter, r *http.Request) {
+	if r.URL.Query().Get("search") != "" {
+		s := r.URL.Query().Get("search")
+		a.SearchTasks(w, r, s)
+	} else {
+		a.GetAllTasks(w)
+	}
+}
+
+func (a *Api) GetAllTasks(w http.ResponseWriter) {
+	foundTasks, err := a.repo.GetAllTasks()
+	if err != nil {
+		log.Println("err:", err)

```

Comment:
Нужно больше в логах писать, а то у тебя в этой функции 3 одинаковых вывода и не понятно будет из какого вызова вылезла ошибка
```suggestion
		log.Println("GetAllTasks err:", err) 
```